stellargraph/layer/rgcn.py

```diff
@@ -0,0 +1,496 @@
+                self.add_weight(
+                    shape=(self.num_bases,),  # hyperparametr B
+                    initializer=self.kernel_initializer,
+                    name="coeff",
+                    regularizer=self.kernel_regularizer,
+                    constraint=self.kernel_constraint,
+                )
+                for _ in range(self.num_relationships)
+            ]
+
+            self.relational_kernels = [
+                tf.einsum("ijk,k->ij", self.bases, coeff) for coeff in self.coefficients
+            ]
+
+        else:
+            self.bases = None
+            self.coefficients = None
+            self.relational_kernels = [
+                self.add_weight(
+                    shape=(input_dim, self.units),
+                    initializer=self.kernel_initializer,
+                    regularizer=self.kernel_regularizer,
+                    constraint=self.kernel_constraint,
+                )
+                for _ in range(self.num_relationships)
+            ]
+
+        self.self_kernel = self.add_weight(
+            shape=(input_dim, self.units),  # hyperparametr B
+            initializer=self.kernel_initializer,
+            regularizer=self.kernel_regularizer,
+            constraint=self.kernel_constraint,
+        )
+
+        if self.use_bias:
+            self.bias = self.add_weight(
+                shape=(self.units,),
+                initializer=self.bias_initializer,
+                name="bias",
+                regularizer=self.bias_regularizer,
+                constraint=self.bias_constraint,
+            )
+        else:
+            self.bias = None
+
+        self.built = True
+
+    def call(self, inputs):
+        """
+        Applies the layer.
+
+        Args:
+            inputs (list): a list of 2 + R input tensors that includes
+                node features (size 1 x N x F),
+                output indices (size 1 x M),
+                and a graph adjacency matrix (size N x N) for each relationship.
+                R is the number of relationships in the graph (edge type),
+                N is the number of nodes in the graph, and
+                F is the dimensionality of node features.
+
+        Returns:
+            Keras Tensor that represents the output of the layer.
+        """
+        features, out_indices, *As = inputs
+        batch_dim, n_nodes, _ = K.int_shape(features)
+        if batch_dim != 1:
+            raise ValueError(
+                "Currently full-batch methods only support a batch dimension of one"
+            )
+
+        # Remove singleton batch dimension
+        features = K.squeeze(features, 0)
+        out_indices = K.squeeze(out_indices, 0)
+
+        # Calculate the layer operation of RGCN
+        output = K.dot(features, self.self_kernel)
+
+        for i in range(self.num_relationships):
+            h_graph = K.dot(As[i], features)
+            output += K.dot(h_graph, self.relational_kernels[i])
+
```

Comment:
Either the above lines are incorrect or I have misunderstood the code or the paper. 

If I am not mistaken, according to equations 2 and 4 in the paper, there is a weight matrix, `W_r`, for each relationship type. In the above code, there is only one W, `self.self_kernel`. Isn't this wrong? In the paper's figure 2, they talk about parameter sharing but I don't think they mean sharing across relationships but rather across nodes. I may be wrong, but can you please check?